pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedConsumerImpl.java

```diff
@@ -418,24 +449,24 @@ private ConsumerConfiguration getInternalConsumerConfig() {
 
     @Override
     public void redeliverUnacknowledgedMessages() {
-        for (ConsumerImpl c : consumers) {
-            c.redeliverUnacknowledgedMessages();
+        synchronized (this) {
+            incomingMessages.clear();
+            unAckedMessageTracker.clear();
+            for (ConsumerImpl c : consumers) {
+                c.redeliverUnacknowledgedMessages();
+            }
+            resumeReceivingFromPausedConsumersIfNeeded();
```

Comment:
what if `pausedConsumers` is empty when it triggers? it will not receive message from any consumer?

Reply:
I'm not sure what you mean.
If `pausedConsumers` is empty that means we're actually waiting for messages from all consumers, this call wouldn't do anything.
The normal flow would lick in, when consumers start getting messages then partitioned will receive messages.

Reply:
I mean what if none of the consumer is waiting on `pausedConsumers`. we add consumer `pausedConsumers` list when `incomingMessages` size reached to threshold. So, if that threshold is not reached yet and `pausedConsumers` is empty and if we call `resumeReceivingFromPausedConsumersIfNeeded` then it doesn't do anything and partitioned-consumer may not receive any messages.

Reply:
@rdhabalia I'm not sure when this could happen.

>  mean what if none of the consumer is waiting on pausedConsumers. we add consumer pausedConsumers list when incomingMessages size reached to threshold

If no consumer is waiting on `pausedConsumers` that would mean we have an active call to `receiveAsync` on every consumer, redelivery should not affect this.

> So, if that threshold is not reached yet and pausedConsumers is empty and if we call resumeReceivingFromPausedConsumersIfNeeded then it doesn't do anything and partitioned-consumer may not receive any messages.

There should be no way this threshold is not reached here since we're emptying the queue, anyways, reaching the threshold would mean that there are still messages in the queue waiting to be consumed, when these messages are eventually consumed that would trigger and unpause of consumers.
Anyways I'm going to reoder the method, I think we should first call redelivery on the consumers and then empty the queue.

Reply:
> If no consumer is waiting on pausedConsumers that would mean we have an active call to receiveAsync on every consumer, redelivery should not affect this.

yes, that's correct.